executor/builder.go

```diff
@@ -103,12 +103,84 @@ func (b *executorBuilder) build(p plan.Plan) Executor {
 		return b.buildUnion(v)
 	case *plan.Update:
 		return b.buildUpdate(v)
+	case *plan.Join:
+		return b.buildJoin(v)
 	default:
 		b.err = ErrUnknownPlan.Gen("Unknown Plan %T", p)
 		return nil
 	}
 }
 
+// compose CNF items into a balance deep CNF tree, which benefits a lot for pb decoder/encoder.
+func composeCondition(conditions []ast.ExprNode) ast.ExprNode {
+	length := len(conditions)
+	if length == 0 {
+		return nil
+	} else if length == 1 {
+		return conditions[0]
+	} else {
+		return &ast.BinaryOperationExpr{Op: opcode.AndAnd, L: composeCondition(conditions[:length/2]), R: composeCondition(conditions[length/2:])}
+	}
+}
+
+//TODO: select join algorithm during cbo phase.
+func (b *executorBuilder) buildJoin(v *plan.Join) Executor {
+	e := &HashJoinExec{
+		otherFilter: composeCondition(v.OtherConditions),
+		prepared:    false,
+		fields:      v.Fields(),
+		ctx:         b.ctx,
+	}
+	var leftHashKey, rightHashKey []ast.ExprNode
+	for _, eqCond := range v.EqualConditions {
+		binop, ok := eqCond.(*ast.BinaryOperationExpr)
+		if ok && binop.Op == opcode.EQ {
+			ln, lOK := binop.L.(*ast.ColumnNameExpr)
+			rn, rOK := binop.R.(*ast.ColumnNameExpr)
+			if lOK && rOK {
+				leftHashKey = append(leftHashKey, ln)
+				rightHashKey = append(rightHashKey, rn)
+				continue
+			}
+		}
+		b.err = ErrUnknownPlan.Gen("Invalid Join Equal Condition !!")
+	}
+	var big, small int
+	switch v.JoinType {
+	case plan.LeftOuterJoin:
+		e.outter = true
+		e.leftSmall = false
+		big, small = 0, 1
```

Comment:
e.leftSmall is enough for deciding which child will be used for smallExec. So big/small is redundant